File: niffler-e-2-e-tests/src/test/java/guru/qa/niffler/page/component/Calendar.java
package guru.qa.niffler.page.component;

import com.codeborne.selenide.ElementsCollection;
import com.codeborne.selenide.Selenide;
import com.codeborne.selenide.SelenideElement;
import io.qameta.allure.Step;

import java.time.Month;
import java.util.Date;
import java.util.GregorianCalendar;

import static com.codeborne.selenide.Selenide.$;
import static java.util.Calendar.DAY_OF_MONTH;
import static java.util.Calendar.MONTH;
import static java.util.Calendar.YEAR;

public class Calendar extends BaseComponent<Calendar> {

  public Calendar(SelenideElement self) {
    super(self);
  }

  public Calendar() {
    super($(".MuiPickersLayout-root"));
  }

  private final SelenideElement input = $("input[name='date']");
  private final SelenideElement calendarButton = $("button[aria-label*='Choose date']");
  private final SelenideElement prevMonthButton = self.$("button[title='Previous month']");
  private final SelenideElement nextMonthButton = self.$("button[title='Next month']");
  private final SelenideElement currentMonthAndYear = self.$(".MuiPickersCalendarHeader-label");
  private final ElementsCollection dateRows = self.$$(".MuiDayCalendar-weekContainer");

  @Step("Select date in calendar: {date}")
  public void selectDateInCalendar(Date date) {
    java.util.Calendar cal = new GregorianCalendar();
    cal.setTime(date);
    calendarButton.click();
    final int desiredMonthIndex = cal.get(MONTH);
    selectYear(cal.get(YEAR));
    selectMonth(desiredMonthIndex);
    selectDay(cal.get(DAY_OF_MONTH));
  }

  private void selectYear(int expectedYear) {
    int actualYear = getActualYear();

    while (actualYear > expectedYear) {
      prevMonthButton.click();
      Selenide.sleep(200);
      actualYear = getActualYear();
    }
    while (actualYear < expectedYear) {
      nextMonthButton.click();
      Selenide.sleep(200);
      actualYear = getActualYear();
    }
  }

  private void selectMonth(int desiredMonthIndex) {
    int actualMonth = getActualMonthIndex();

    while (actualMonth > desiredMonthIndex) {
      prevMonthButton.click();
      Selenide.sleep(200);
      actualMonth = getActualMonthIndex();
    }
    while (actualMonth < desiredMonthIndex) {
      nextMonthButton.click();
      Selenide.sleep(200);
      actualMonth = getActualMonthIndex();
    }
  }

  private void selectDay(int desiredDay) {
    ElementsCollection rows = dateRows.snapshot();

    for (SelenideElement row : rows) {
      ElementsCollection days = row.$$("button").snapshot();
      for (SelenideElement day : days) {
        if (day.getText().equals(String.valueOf(desiredDay))) {
          day.click();
          return;
        }
      }
    }
  }

  private String getMonthNameByIndex(int mothIndex) {
    return Month.of(mothIndex + 1).name();
  }

  private int getActualMonthIndex() {
    return Month.valueOf(currentMonthAndYear.getText()
            .split(" ")[0]
            .toUpperCase())
        .ordinal();
  }

  private int getActualYear() {
    return Integer.parseInt(currentMonthAndYear.getText().split(" ")[1]);
  }
}

File: niffler-e-2-e-tests-python/pages/components/calendar.py
"""Calendar component for date selection in Niffler application"""

from selene import browser, be, have
import allure
from datetime import date
import time
from calendar import month_name

from pages.components.base_component import BaseComponent


class Calendar(BaseComponent):
    """Calendar component matching Java Calendar class"""

    DATE_REGEXP = r"^\w+ \d{4}$"

    def __init__(self, root=None):
        """
        Initialize calendar component

        Args:
            root: Root element of calendar. If None, uses default selector
        """
        if root is None:
            root = browser.element('.MuiPickersLayout-root')
        super().__init__(root)

        self._input = browser.element('input[name="date"]')
        self._calendar_button = browser.element('button[aria-label*="Choose date"]')
        self._prev_month_button = self.self.element('button[title="Previous month"]')
        self._next_month_button = self.self.element('button[title="Next month"]')
        self._current_month_and_year = self.self.element('.MuiPickersCalendarHeader-label')
        self._date_rows = self.self.all('.MuiDayCalendar-weekContainer')

    @allure.step("Select date in calendar: {date_to_select}")
    def select_date_in_calendar(self, date_to_select: date) -> 'Calendar':
        """
        Select a date in the calendar picker

        Args:
            date_to_select: Python date object to select

        Returns:
            self for method chaining
        """
        self._calendar_button.click()

        desired_year = date_to_select.year
        desired_month_index = date_to_select.month - 1  # 0-indexed like Java Calendar
        desired_day = date_to_select.day

        self._select_year(desired_year)
        self._select_month(desired_month_index)
        self._select_day(desired_day)

        return self

    def _select_year(self, expected_year: int):
        """Navigate to the correct year"""
        actual_year = self._get_actual_year()

        while actual_year > expected_year:
            self._prev_month_button.click()
            time.sleep(0.2)
            actual_year = self._get_actual_year()

        while actual_year < expected_year:
            self._next_month_button.click()
            time.sleep(0.2)
            actual_year = self._get_actual_year()

    def _select_month(self, desired_month_index: int):
        """Navigate to the correct month (0-indexed)"""
        actual_month = self._get_actual_month_index()

        while actual_month > desired_month_index:
            self._prev_month_button.click()
            time.sleep(0.2)
            actual_month = self._get_actual_month_index()

        while actual_month < desired_month_index:
            self._next_month_button.click()
            time.sleep(0.2)
            actual_month = self._get_actual_month_index()

    def _select_day(self, desired_day: int):
        """Click on the desired day in the calendar grid"""
        rows = self._date_rows

        for row in rows:
            days = row.all('button')
            for day_button in days:
                if day_button.get(be.visible).text == str(desired_day):
                    day_button.click()
                    return

    def _get_actual_month_index(self) -> int:
        """
        Get current month index from calendar header (0-indexed)

        Returns:
            Month index (0 = January, 11 = December)
        """
        self._current_month_and_year.should(have.text_matching(self.DATE_REGEXP))
        month_text = self._current_month_and_year.get(be.visible).text.split()[0]

        # Convert month name to index
        for i, name in enumerate(month_name):
            if name.upper() == month_text.upper():
                return i - 1  # month_name starts with empty string at index 0

        raise ValueError(f"Could not parse month from: {month_text}")

    def _get_actual_year(self) -> int:
        """
        Get current year from calendar header

        Returns:
            Year as integer
        """
        self._current_month_and_year.should(have.text_matching(self.DATE_REGEXP))
        year_text = self._current_month_and_year.get(be.visible).text.split()[1]
        return int(year_text)
